homeassistant/components/essent/coordinator.py

```diff
@@ -0,0 +1,177 @@
+        _LOGGER.info(
+            "Starting schedules: API fetch every hour at minute %d, "
+            "listener updates on the hour",
+            self._api_fetch_minute_offset,
+        )
+        self._schedule_data_refresh()
+        self._schedule_listener_tick()
+
+    async def async_shutdown(self) -> None:
+        """Cancel any scheduled call, and ignore new runs."""
+        await super().async_shutdown()
+        if self._unsub_data:
+            self._unsub_data()
+            self._unsub_data = None
+        if self._unsub_listener:
+            self._unsub_listener()
+            self._unsub_listener = None
+
+    def _schedule_data_refresh(self) -> None:
+        """Schedule next data fetch at a random minute offset within the hour."""
+        if self._unsub_data:
+            self._unsub_data()
+
+        now = dt_util.utcnow()
+        current_hour = now.replace(minute=0, second=0, microsecond=0)
+        candidate = (
+            current_hour
+            + UPDATE_INTERVAL
+            + timedelta(minutes=self._api_fetch_minute_offset)
+        )
+        if candidate <= now:
+            candidate = candidate + UPDATE_INTERVAL
+
+        _LOGGER.debug(
+            "Scheduling next API fetch for %s (offset: %d minutes)",
+            candidate,
+            self._api_fetch_minute_offset,
+        )
+
+        @callback
+        def _handle(_: datetime) -> None:
+            """Handle the scheduled API refresh trigger."""
+            self._unsub_data = None
+            self.hass.async_create_task(self.async_request_refresh())
+            # Reschedule for next hour regardless of success/failure
+            self._schedule_data_refresh()
+
+        self._unsub_data = async_track_point_in_utc_time(self.hass, _handle, candidate)
+
+    def _schedule_listener_tick(self) -> None:
+        """Schedule listener updates on the hour to advance cached tariffs."""
+        if self._unsub_listener:
+            self._unsub_listener()
+
+        now = dt_util.utcnow()
+        next_hour = now + UPDATE_INTERVAL
+        next_run = datetime(
+            next_hour.year,
+            next_hour.month,
+            next_hour.day,
+            next_hour.hour,
+            tzinfo=dt_util.UTC,
+        )
+
+        _LOGGER.debug("Scheduling next listener tick for %s", next_run)
+
+        @callback
+        def _handle(_: datetime) -> None:
+            """Handle the scheduled listener tick to update sensors."""
+            self._unsub_listener = None
+            _LOGGER.debug("Listener tick fired, updating sensors with cached data")
+            self.async_update_listeners()
+            self._schedule_listener_tick()
+
+        self._unsub_listener = async_track_point_in_utc_time(
+            self.hass,
+            _handle,
+            next_run,
+        )
+
+    async def _async_update_data(self) -> EssentData:
+        """Fetch data from API."""
+        try:
+            prices: EssentPrices = await self._client.async_get_prices()
```

Comment:
Fyi, mypy can't read the types of your library until you add a py.typed file to the root of the library, so if your library is well typed you can add that

Reply:
yes! Good call, added this to the library.